src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -11,68 +11,345 @@
-            throw new \LogicException(\sprintf('Using "%s" requires the HttpKernel component, try running "composer require symfony/http-kernel".', __CLASS__));
-        }
+        if ($cache instanceof StoreInterface) {
+            trigger_deprecation('symfony/http-client', '7.3', 'Passing a "%s" as constructor\'s 2nd argument of "%s" is deprecated, "%s" expected.', StoreInterface::class, __CLASS__, TagAwareAdapterInterface::class);
 
-        $kernel = new HttpClientKernel($client);
-        $this->cache = new HttpCache($kernel, $store, null, $defaultOptions);
+            if (!class_exists(HttpClientKernel::class)) {
+                throw new \LogicException(\sprintf('Using "%s" requires the HttpKernel component, try running "composer require symfony/http-kernel".', __CLASS__));
+            }
+
+            $kernel = new HttpClientKernel($client);
+            $this->cache = new HttpCache($kernel, $cache, null, $defaultOptions);
 
-        unset($defaultOptions['debug']);
-        unset($defaultOptions['default_ttl']);
-        unset($defaultOptions['private_headers']);
-        unset($defaultOptions['skip_response_headers']);
-        unset($defaultOptions['allow_reload']);
-        unset($defaultOptions['allow_revalidate']);
-        unset($defaultOptions['stale_while_revalidate']);
-        unset($defaultOptions['stale_if_error']);
-        unset($defaultOptions['trace_level']);
-        unset($defaultOptions['trace_header']);
+            unset($defaultOptions['debug']);
+            unset($defaultOptions['default_ttl']);
+            unset($defaultOptions['private_headers']);
+            unset($defaultOptions['skip_response_headers']);
+            unset($defaultOptions['allow_reload']);
+            unset($defaultOptions['allow_revalidate']);
+            unset($defaultOptions['stale_while_revalidate']);
+            unset($defaultOptions['stale_if_error']);
+            unset($defaultOptions['trace_level']);
+            unset($defaultOptions['trace_header']);
+        } else {
+            $this->cache = $cache;
+        }
 
         if ($defaultOptions) {
             [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, $this->defaultOptions);
         }
     }
 
     public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->cache instanceof HttpCache) {
+            return $this->legacyRequest($method, $url, $options);
+        }
+
+        [$fullUrl, $options] = self::prepareRequest($method, $url, $options, $this->defaultOptions);
+
+        $fullUrl = implode('', $fullUrl);
+        $fullUrlTag = hash('xxh3', $fullUrl);
+
+        if (\in_array($method, self::UNSAFE_METHODS, true)) {
+            $this->cache->invalidateTags([$fullUrlTag]);
+        }
+
+        if (!empty($options['body']) || isset($options['normalized_headers']['range']) || !\in_array($method, self::CACHEABLE_METHODS, true)) {
```

Comment:
```suggestion
        if ($options['body'] || isset($options['normalized_headers']['range']) || !\in_array($method, self::CACHEABLE_METHODS, true)) {
```

`empty()` is kind of banned from the codebase, `$options['body']` should be sufficient?

Reply:
adressed in https://github.com/symfony/symfony/pull/59576/commits/ee173e8a09a358467ee302c9443259d047f551e8